### ISSUE TITLE
Unable to get file association working

### ISSUE THREAD

**Issue by balaji-dutt** [Opticos/OpenInWSL-Source#15]
Hi,

I'm trying to open `.org` mode files using programs installed in WSL2 on Windows 10. I have tried this with both `emacs` and then for debugging with `gedit` but this does not seem to work. I have also tried switching the Graphical Backend between `GWSL` and `wslg` but there's no change.

Here is what I'm doing:

- Set file association for `.org` in OpeninWSL to gedit:
![image](https://user-images.githubusercontent.com/3147070/230537902-6afc6ea3-8ac1-418f-9849-4f70975b0adc.png)

- Set file association for `.org` file in Windows Explorer to "OpeninWSL"
![image](https://user-images.githubusercontent.com/3147070/230538073-f3dd2f2e-c275-4e7d-9cb4-2f7354856fe8.png)

- Double click on a file `hobbies.org` in Windows Explorer.
- What I expect to see
  - A `gedit` window will open with the file.
- What happens instead:
  - I see the popup animation showing ".org >>> gedit" but nothing happens afterwards.

- Logs:
~~~
2023-04-07 11:43:12,112 - OpenInWSL 1.7 MSIX - INFO - RUN TEST: emacs28 '/path/to/projects/hobbies.org'
2023-04-07 11:43:56,570 - OpenInWSL 1.7 MSIX - INFO - RUN TEST: emacs28 '/path/to/projects/hobbies.org'
2023-04-07 11:44:45,612 - OpenInWSL 1.7 MSIX - INFO - RUN TEST: gedit --new-window '/path/to/projects/hobbies.org'
2023-04-07 11:46:10,243 - OpenInWSL 1.7 MSIX - INFO - RUN TEST: gedit '/path/to/projects/hobbies.org'
2023-04-07 11:51:14,586 - OpenInWSL 1.7 MSIX - INFO - RUN TEST: gedit '/path/to/projects/hobbies.org'
~~~

<details>
  <summary>Version Info</summary>
 Ubuntu 22.04 running in WSL2/Windows 10
Open in WSL: 1.7 MSIX
GWSL: 1.4.5 (Store)
</details>

**Comment by Pololot64** [comment]
What happens if you run the commands mentioned in the logs directly eg ```emacs28 '/path/to/projects/hobbies.org'```

Does anything run? Is an xserver or window server running correctly? This is a prerequisite

**Comment by balaji-dutt** [comment]
> What happens if you run the commands mentioned in the logs directly eg `emacs28 '/path/to/projects/hobbies.org'`
> 

This works correctly if I run the command from a WSL2 terminal - see screenshots below:
![WindowsTerminal_eVV0OgamO7](https://user-images.githubusercontent.com/3147070/230700636-b6f357df-4f1d-4d99-86b2-deefea0b6cd5.png)
![msrdc_RjvGSKrIk3](https://user-images.githubusercontent.com/3147070/230700639-605ac810-0788-46e0-8907-b418f94295b0.png)

> Does anything run? Is an xserver or window server running correctly? This is a prerequisite

I have setup Open in WSL to use `wslg`. I have tested opening an `.org` file with "Open in WSL" while having a WSL2 terminal open but that does not seem to  make a difference. 'GWSL` was throwing some errors so I wasn't able to try that option. I tried to open an `.org` file with VcXsrv running but that did not work either.

**Comment by balaji-dutt** [comment]
> What happens if you run the commands mentioned in the logs directly eg `emacs28 '/path/to/projects/hobbies.org'`
> 
![WindowsTerminal_eVV0OgamO7](https://user-images.githubusercontent.com/3147070/230700837-fce258df-7281-4c09-bb47-a4f5fd32f177.png)
![msrdc_RjvGSKrIk3](https://user-images.githubusercontent.com/3147070/230700841-06d1d40d-213f-404c-a9e7-7f5f96c557e3.png)



This works correctly if I run the command from a WSL2 terminal - see screenshots below:

> Does anything run? Is an xserver or window server running correctly? This is a prerequisite

I have setup Open in WSL to use `wslg`. I have tested opening an `.org` file with "Open in WSL" while having a WSL2 terminal open but that does not seem to  make a difference. `GWSL` was throwing some errors so I wasn't able to try that option. I tried to open an `.org` file with `VcXsrv` running but that did not work either.

**Comment by Pololot64** [comment]
Does any other file type work?

**Comment by balaji-dutt** [comment]
> Does any other file type work?

Things seem to be getting worse :-(.

I tried to set "Open in WSL" as the default program for `.patch` first and then for a dummy extension `.ext`. I created `test.ext` using the following command `$ echo "foo" > test.ext`.

For both `.patch` and `.ext` I'm seeing the following error:

~~~sh
2023-04-08 12:52:44,936 - OpenInWSL 1.7 MSIX - ERROR - Cannot Open extension error
Traceback (most recent call last):
  File "main.py", line 2272, in <module>
  File "main.py", line 515, in path_converter
IndexError: string index out of range
2023-04-08 12:53:14,075 - OpenInWSL 1.7 MSIX - ERROR - Cannot Open extension error
Traceback (most recent call last):
  File "main.py", line 2272, in <module>
  File "main.py", line 515, in path_converter
IndexError: string index out of range
2023-04-08 12:55:39,588 - OpenInWSL 1.7 MSIX - ERROR - Cannot Open extension error
Traceback (most recent call last):
  File "main.py", line 2272, in <module>
  File "main.py", line 515, in path_converter
IndexError: string index out of range
~~~

I tried to uninstall & reinstall Open in WSL via MS-Store but that does not seem to help as the settings/logs etc seem to persist between uninstalls.

**Comment by Pololot64** [comment]
Ok we are getting somewhere. There is probably an issue with openinwsl parsing the windows path. Could you post the path to the file censoring only the words but not the structure (eg spaces, slashes, drive, punctuation)?

**Comment by balaji-dutt** [comment]
> Ok we are getting somewhere. There is probably an issue with openinwsl parsing the windows path. Could you post the path to the file censoring only the words but not the structure (eg spaces, slashes, drive, punctuation)?

Apologies I realized later that the `string index out of range` error was due to my error. I was trying to open a file in the path `\\wsl.localhost\Ubuntu-22.04\home\balaji\src\test.ext`.

However I do have some good news!

I copied the file `test.ext` to the path `C:\batch\test.ext` and when I double clicked on the file in this path, it opened correctly in `gedit`. I also tested `.org` files by copying `hobbies.org` to `C:\batch\hobbies.org` and that worked perfectly with `emacs28`.

So indeed it appears that there's an issue with parsing the windows path in openinwsl.

Here is a lightly censored path to the original location of the `hobbies.org` file `/mnt/c/balaji/xxxxxxx/notes & projects/projects/hobbies.org`. As before this fails without any real error in the logs

~~~
2023-04-09 16:31:05,413 - OpenInWSL 1.7 MSIX - INFO - RUN TEST: emacs28 '/mnt/f/balaji/xxxxxxx/notes & projects/projects/hobbies.org'
~~~

**Comment by balaji-dutt** [comment]
Hi @Pololot64 I was wondering if you had any suggestions / updates on this issue?

**Comment by Pololot64** [comment]
Not yet. I have been doing exams. Well, I am a little confused about what the problem still is. I see that when the path is in Windows, it works properly. Is it only breaking when you use wsl paths like \\wsl.localhost ?

**Comment by balaji-dutt** [comment]
Apologies if I've been unclear before. I will try and re-explain my current status:

- I'm trying to open the file `c:\balaji\xxxxxxx\notes & projects\projects\hobbies.org` from Windows Explorer by double-clicking on the file.
  - `.org` files are set to open with `OpenIn WSL` in Windows.
  - In `OpenIn WSL`, `.org` files are configured to open with `emacs`

- When I double-click the file  `c:\balaji\xxxxxxx\notes & projects\projects\hobbies.org`, I see the ".org >>>>" pop-up in Windows, but the file does not open.
  - The only message in the `OpenIn WSL` log file is as follows:
  ```
  2023-04-09 16:31:05,413 - OpenInWSL 1.7 MSIX - INFO - RUN TEST: emacs28 '/mnt/c/balaji/xxxxxxx/notes & projects/projects/hobbies.org'
  ```

- If I copy the file `c:\balaji\xxxxxxx\notes & projects\projects\hobbies.org` to a different path with no spaces in the path, for example, `c:\batch\hobbies.org`, an `emacs` window does appear with the file open as expected. 
  - The message in the log file is as follows:
  ```
  2023-04-09 16:28:04,428 - OpenInWSL 1.7 MSIX - INFO - RUN TEST: emacs28 /mnt/c/batch/hobbies.org
  ```
  - In other words, the messages in the log file are identical whether the program appears with the file or nothing happens at all.

- The only time I saw an actual error message was when I tried to open a file from `\\wsl.localhost\Ubuntu-22.04\home\balaji\src\test.ext` in Windows Explorer. I'm not sure that this type of UNC path is supported and hence would believe this particular error is not relevant to the problem mentioned above.

**Comment by balaji-dutt** [comment]
> I see. In that case my second theory is that it is the "&" symbol in the path

Is that something that can be supported through an enhancement to "OpenIn WSL"?

**Comment by balaji-dutt** [comment]
Ah I see. Well this is pretty disappointing news ☹️

**Comment by Pololot64** [comment]
Again however, if this is extremely important feel free to email me and I can try to build you a version that fixes this. Also, theoretically when I am done with school I can make it try to escape everything but that will be hard to predict and plan for.

**Comment by balaji-dutt** [comment]
> Again however, if this is extremely important feel free to email me and I can try to build you a version that fixes this. Also, theoretically when I am done with school I can make it try to escape everything but that will be hard to predict and plan for.

Hi @Pololot64 I have emailed you on your `opticosstudios` address on this.

**Comment by Pololot64** [comment]
I'll get back to you
